flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java

```diff
@@ -0,0 +1,238 @@
+        if (request.getNumberOfRetries() > 0) {
+          restored.add(request);
+          dataFilesNumRestored = dataFilesNumRestored + committable.dataFiles().length;
+          deleteFilesNumRestored = deleteFilesNumRestored + committable.deleteFiles().length;
+        } else {
+          store.add(request);
+          dataFilesNumStored = dataFilesNumStored + committable.dataFiles().length;
+          deleteFilesNumStored = deleteFilesNumStored + committable.deleteFiles().length;
+        }
+        checkpointId = Math.max(checkpointId, request.getCommittable().checkpointId());
+      }
+    }
+
+    if (restored.size() > 0) {
+      commitResult(dataFilesNumRestored, deleteFilesNumRestored, restored);
+    }
+
+    commitResult(dataFilesNumStored, deleteFilesNumStored, store);
+    this.maxCommittedCheckpointId = checkpointId;
+  }
+
+  @Override
+  public void close() throws Exception {
+    if (tableLoader != null) {
+      tableLoader.close();
+    }
+
+    if (workerPool != null) {
+      workerPool.shutdown();
+    }
+  }
+
+  private void commitResult(int dataFilesNum, int deleteFilesNum,
+                            Collection<CommitRequest<FilesCommittable>> committableCollection) {
+    int totalFiles = dataFilesNum + deleteFilesNum;
+
+    if (totalFiles != 0) {
+      if (replacePartitions) {
+        replacePartitions(committableCollection, dataFilesNum, deleteFilesNum);
+      } else {
+        commitDeltaTxn(committableCollection, dataFilesNum, deleteFilesNum);
+      }
+    }
+  }
+
+  private void replacePartitions(Collection<CommitRequest<FilesCommittable>> writeResults, int dataFilesNum,
+                                 int deleteFilesNum) {
+    // Partition overwrite does not support delete files.
+    Preconditions.checkState(deleteFilesNum == 0, "Cannot overwrite partitions with delete files.");
+
+    // Commit the overwrite transaction.
+    ReplacePartitions dynamicOverwrite = table.newReplacePartitions().scanManifestsWith(workerPool);
+
+    String flinkJobId = null;
+    long checkpointId = -1;
+    for (CommitRequest<FilesCommittable> commitRequest : writeResults) {
+      try {
+        WriteResult result = commitRequest.getCommittable().committable();
+        Preconditions.checkState(result.referencedDataFiles().length == 0, "Should have no referenced data files.");
+
+        flinkJobId = commitRequest.getCommittable().jobID();
+        checkpointId = commitRequest.getCommittable().checkpointId();
+        Arrays.stream(result.dataFiles()).forEach(dynamicOverwrite::addFile);
+      } catch (Exception e) {
+        LOG.error("Unable to process the committable: {}.", commitRequest.getCommittable(), e);
+        commitRequest.signalFailedWithUnknownReason(e);
+      }
+    }
+
+    commitOperation(dynamicOverwrite, dataFilesNum, 0, "dynamic partition overwrite", flinkJobId, checkpointId);
+    writeResults.forEach(CommitRequest::signalAlreadyCommitted);
```

Comment:
I still think it is not correct to call this method here, since based on the documentation it will write logs that we tried to recommit the same change again (even if this is only the first time when we try to commit this specific change)

Reply:
I've removed it here and will use it only in the case of an exception that still thinks it's committed.